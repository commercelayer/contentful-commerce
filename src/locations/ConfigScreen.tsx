import React, { useCallback, useState, useEffect } from "react";
import { ConfigAppSDK, ContentType } from "@contentful/app-sdk";
import {
  Form,
  Paragraph,
  Flex,
  TextInput,
  FormControl,
} from "@contentful/f36-components";
import { css } from "@emotion/css";
import { useSDK } from "@contentful/react-apps-toolkit";
import { styles } from "./ConfigScreen.styles";
import getSchema, { ContentTypeConfiguration } from "../schema";

const createAndPublishContentType = async (
  sdk: ConfigAppSDK,
  { fields, ...contentType }: ContentTypeConfiguration,
): Promise<void> => {
  try {
    const { sys } = await sdk.cma.contentType.createWithId(
      { contentTypeId: contentType.id },
      { fields, ...contentType },
    );

    await sdk.cma.contentType.publish(
      { contentTypeId: sys.id },
      {
        sys,
        ...contentType,
        fields: [],
      },
    );

    console.log(`Content type ${contentType.name} created and published.`);
  } catch (error) {
    console.error(`Error creating content type ${contentType.name}:`, error);
  }
};

const updateContentType = (
  sdk: ConfigAppSDK,
  { name, ...contentType }: ContentType,
  label: string,
) => {
  console.log("updating", name, label);
  if (label === name) {
    console.log("skip for " + name);
    return;
  }

  return sdk.cma.contentType.update(
    { contentTypeId: contentType.sys.id },
    {
      name: label,
      ...contentType,
    },
  );
};

export interface AppInstallationParameters {
  productLabel?: string;
  variantLabel?: string;
  taxonomyLabel?: string;
  taxonLabel?: string;
}

const configuration: {
  label: string;
  name: keyof AppInstallationParameters;
  isRequired: boolean;
  helpText: string;
  placeholder: string;
}[] = [
  {
    label: "Product Label",
    name: "productLabel",
    isRequired: false,
    helpText: "The label of the Product content type",
    placeholder: "e.g. Book",
  },
  {
    label: "Variant Label",
    name: "variantLabel",
    isRequired: false,
    helpText: "The label of the Variant content type",
    placeholder: "e.g. Book Type",
  },
  {
    label: "Taxonomy Label",
    name: "taxonomyLabel",
    isRequired: false,
    helpText: "The label of the Taxonomy content type",
    placeholder: "e.g. Category System",
  },
  {
    label: "Taxon Label",
    name: "taxonLabel",
    isRequired: false,
    helpText: "The label of the Taxon content type",
    placeholder: "e.g. Category",
  },
];

const ConfigScreen = () => {
  const [parameters, setParameters] = useState<AppInstallationParameters>({});
  const sdk = useSDK<ConfigAppSDK>();

  console.log("Config screen rendered.");

  const onConfigure = useCallback(async () => {
    const currentState = await sdk.app.getCurrentState();
    const schema = getSchema({ localized: false, ...parameters });

    for (const contentType of schema) {
      const existingContentType = await sdk.cma.contentType.get({
        contentTypeId: contentType.id,
      });

      if (existingContentType) {
        await updateContentType(sdk, existingContentType, contentType.name);
      } else {
        await createAndPublishContentType(sdk, contentType);
      }
    }

    return {
      parameters,
      targetState: currentState,
    };
  }, [parameters, sdk]);

  useEffect(() => {
    sdk.app.onConfigure(() => onConfigure());
  }, [sdk, onConfigure]);

  useEffect(() => {
    (async () => {
      // Get current parameters of the app.
      // If the app is not installed yet, `parameters` will be `null`.
      const currentParameters: AppInstallationParameters | null =
        await sdk.app.getParameters();

      if (currentParameters) {
        setParameters(currentParameters);
      }

      // Once preparation has finished, call `setReady` to hide
      // the loading screen and present the app to a user.
      sdk.app.setReady();
    })();
  }, [sdk]);

  return (
    <Flex
      flexDirection="column"
      className={css({ margin: "80px", maxWidth: "800px" })}
    >
      <div className={styles.background}>
        <div className={styles.body}>
          <img
            src="assets/header-light.png"
            alt="Contentful Commerce by Commerce Layer"
          />
          <div className={styles.content}>
            <Paragraph>
              Welcome to the Configuration page of the Contentful Commerce app,
              powered by Commerce Layer. This interface is designed to
              streamline your journey in setting up a product catalog within
              Contentful. Tailor your content model to fit your unique commerce
              needs by customizing content type labels, which will be reflected
              across your Contentful environment.
            </Paragraph>
            <Paragraph>
              Please note that while you can personalize the display labels, the
              underlying ID used in the API responses will remain consistent,
              ensuring a stable and seamless integration. Start configuring your
              content types below to enhance your ecommerce experience.
            </Paragraph>
            <hr className={styles.splitter} />
            <Form>
              {configuration.map((field) => {
                return (
                  <FormControl isInvalid={false}>
                    <FormControl.Label>{field.label}</FormControl.Label>
                    <TextInput
                      name={field.name}
                      id={field.name}
                      placeholder={field.placeholder}
                      value={parameters[field.name] || ""}
                      onChange={(e) => {
                        setParameters({
                          ...parameters,
                          [field.name]: e.target.value,
                        });
                      }}
                    />
                    <FormControl.HelpText>
                      {field.helpText}
                    </FormControl.HelpText>
                  </FormControl>
                );
              })}
            </Form>
          </div>
        </div>
        <div className={styles.icon}>
          <img
            src="assets/commercelayer-glyph-black.png"
            alt="Commerce Layer logo"
          />
        </div>
      </div>
    </Flex>
  );
};

export default ConfigScreen;
