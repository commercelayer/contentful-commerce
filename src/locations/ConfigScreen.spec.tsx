import React from "react";
import ConfigScreen from "./ConfigScreen";
import { render } from "@testing-library/react";
import { mockCma, mockSdk } from "../../test/mocks";

jest.mock("@contentful/react-apps-toolkit", () => ({
  useSDK: () => mockSdk,
  useCMA: () => mockCma,
}));

describe("Config Screen component", () => {
  it("renders correctly", () => {
    const { container } = render(<ConfigScreen />);
    const dynamicAttributeElements = container.querySelectorAll("[id],[for]");

    for (const element of Array.from(dynamicAttributeElements)) {
      element.removeAttribute("id");
      element.removeAttribute("for");
    }

    expect(container.firstChild).toMatchSnapshot();
  });
});
