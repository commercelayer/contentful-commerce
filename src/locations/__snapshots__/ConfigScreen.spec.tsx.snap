// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Screen component renders correctly 1`] = `
<div
  class="css-5mda1q css-j7qwjs"
>
  <div
    class="css-1ydpq0v"
  >
    <div
      class="css-1rggvfa"
    >
      <img
        alt="Contentful Commerce by Commerce Layer"
        src="assets/header-light.png"
      />
      <div
        class="css-14u56od"
      >
        <p
          class="css-ki4iwd"
          data-test-id="cf-ui-paragraph"
        >
          Welcome to the Configuration page of the Contentful Commerce app, powered by Commerce Layer. This interface is designed to streamline your journey in setting up a product catalog within Contentful. Tailor your content model to fit your unique commerce needs by customizing content type labels, which will be reflected across your Contentful environment.
        </p>
        <p
          class="css-ki4iwd"
          data-test-id="cf-ui-paragraph"
        >
          Please note that while you can personalize the display labels, the underlying ID used in the API responses will remain consistent, ensuring a stable and seamless integration. Start configuring your content types below to enhance your ecommerce experience.
        </p>
        <hr
          class="css-10hsrou"
        />
        <form
          data-test-id="cf-ui-form-label"
        >
          <div
            class="css-1fobf8d"
            data-test-id="cf-ui-form-control"
            role="group"
          >
            <label
              class="css-1ry5edl"
              data-test-id="cf-ui-form-label"
            >
              Product Label
            </label>
            <input
              aria-describedby="productLabel-helptext"
              class="css-1ml10qx"
              data-test-id="cf-ui-text-input"
              name="productLabel"
              placeholder="e.g. Book"
              type="text"
              value=""
            />
            <p
              class="css-1l2c7z"
              data-test-id="cf-ui-help-text"
            >
              The label of the Product content type
            </p>
          </div>
          <div
            class="css-1fobf8d"
            data-test-id="cf-ui-form-control"
            role="group"
          >
            <label
              class="css-1ry5edl"
              data-test-id="cf-ui-form-label"
            >
              Variant Label
            </label>
            <input
              aria-describedby="variantLabel-helptext"
              class="css-1ml10qx"
              data-test-id="cf-ui-text-input"
              name="variantLabel"
              placeholder="e.g. Book Type"
              type="text"
              value=""
            />
            <p
              class="css-1l2c7z"
              data-test-id="cf-ui-help-text"
            >
              The label of the Variant content type
            </p>
          </div>
          <div
            class="css-1fobf8d"
            data-test-id="cf-ui-form-control"
            role="group"
          >
            <label
              class="css-1ry5edl"
              data-test-id="cf-ui-form-label"
            >
              Taxonomy Label
            </label>
            <input
              aria-describedby="taxonomyLabel-helptext"
              class="css-1ml10qx"
              data-test-id="cf-ui-text-input"
              name="taxonomyLabel"
              placeholder="e.g. Category System"
              type="text"
              value=""
            />
            <p
              class="css-1l2c7z"
              data-test-id="cf-ui-help-text"
            >
              The label of the Taxonomy content type
            </p>
          </div>
          <div
            class="css-1fobf8d"
            data-test-id="cf-ui-form-control"
            role="group"
          >
            <label
              class="css-1ry5edl"
              data-test-id="cf-ui-form-label"
            >
              Taxon Label
            </label>
            <input
              aria-describedby="taxonLabel-helptext"
              class="css-1ml10qx"
              data-test-id="cf-ui-text-input"
              name="taxonLabel"
              placeholder="e.g. Category"
              type="text"
              value=""
            />
            <p
              class="css-1l2c7z"
              data-test-id="cf-ui-help-text"
            >
              The label of the Taxon content type
            </p>
          </div>
        </form>
      </div>
    </div>
    <div
      class="css-1cypcly"
    >
      <img
        alt="Commerce Layer logo"
        src="assets/commercelayer-glyph-black.png"
      />
    </div>
  </div>
</div>
`;
